File: src/application/Application.java
package application;

import OrganizationDatabase.Organization;
import OrganizationDatabase.OrganizationDatabase;
import OrganizationDatabase.OrganizationManagerInterface;
import commands.client_side.core.Command;
import lib.BufferedReaderWithQueueOfStreams;
import lib.ExecutionStatus;
import lib.Localization;
import lib.collections.CircledStorage;
import lib.collections.ImmutablePair;
import server.DatabaseCommandsReceiver;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class Application {
    private final OrganizationManagerInterface organizationDatabase = new OrganizationDatabase();
    private final CircledStorage<String> commandsHistory = new CircledStorage<>(11);
    private final BufferedReaderWithQueueOfStreams bufferedReaderWithQueueOfStreams =
            new BufferedReaderWithQueueOfStreams(new InputStreamReader(System.in));

    private boolean needToStop = false;
    private HashMap<String, Command> commandsWithLocaleNames;

    private final List<ImmutablePair<String, Command>> commandList = ApplicationCommandInitializer.getCommand(this);

    public BufferedReaderWithQueueOfStreams getBufferedReaderWithQueueOfStreams() {
        return bufferedReaderWithQueueOfStreams;
    }

    public CircledStorage<String> getCommandsHistory() {
        return commandsHistory;
    }

    public OrganizationManagerInterface getOrganizationManager() {
        return organizationDatabase;
    }

    public void start(String database) throws IOException {
        if (database != null) {
            organizationDatabase.loadFromFile(database);
        }

        DatabaseCommandsReceiver databaseCommandsReceiver = new DatabaseCommandsReceiver(6789);
        databaseCommandsReceiver.run();

        localize();
        printIntroductionMessage();

        while (!needToStop) {
            processCommand(bufferedReaderWithQueueOfStreams.readLine().strip());
        }
    }

    public void stop() {
        needToStop = true;
    }

    private void printCommandResult(ExecutionStatus status, String successMessage, String failureMessage) {
        if (status == ExecutionStatus.SUCCESS) {
            System.out.println(successMessage);
        } else {
            System.out.println(failureMessage);
        }
    }

    private void printCommandResult(ExecutionStatus status, String successMessage) {
        printCommandResult(status, successMessage, Localization.get("message.command.failed"));
    }

    void exitCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 0;

        printCommandResult(
                status,
                Localization.get("message.exit")
        );
    }

    void showCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert (result.length == 1 && result[0] instanceof String) || error != null;

        printCommandResult(
                status,
                (String) result[0],
                ApplicationErrorMessages.showCommandGetErrorMessage(error)
        );
    }

    void helpCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        showCommandCallback(status, error, result); // info, help and show commands have the same callback
    }

    void infoCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        showCommandCallback(status, error, result); // info, help and show commands have the same callback
    }

    void clearCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 0;

        printCommandResult(
                status,
                Localization.get("message.collection_cleared")
        );
    }

    void saveCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert (result.length == 1 && result[0] instanceof String) || error != null;

        String filename = (String) result[0];
        printCommandResult(
                status,
                String.format("%s %s.", Localization.get("message.collection.saved_to_file"), filename),
                String.format("%s %s.", Localization.get("message.collection.unable_to_save_to_file"), filename)
        );
    }

    void readCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert (result.length == 1 && result[0] instanceof String) || error != null;

        String filename = (String) result[0];

        printCommandResult(
                status,
                String.format("%s.", Localization.get("message.collection.load.succeed")),
                String.format("%s %s.", Localization.get("message.collection.load.failed"), filename)
        );
    }

    void addCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 0;

        printCommandResult(
                status,
                Localization.get("message.collection.add.succeed"),
                ApplicationErrorMessages.addCommandGetErrorMessage(error)
        );
    }

    void addMaxCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 0;

        printCommandResult(
                status,
                Localization.get("message.collection.add.succeed"),
                ApplicationErrorMessages.addMaxCommandGetErrorMessage(error)
        );
    }

    void removeCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 0;

        printCommandResult(
                status,
                Localization.get("message.organization_removed"),
                ApplicationErrorMessages.removeByIdCommandGetErrorMessage(error)
        );
    }

    void removeHeadCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert (result.length == 1 && result[0] instanceof String) || error != null;

        printCommandResult(
                status,
                (String) result[0],
                ApplicationErrorMessages.removeHeadCommandGetErrorMessage(error)
        );
    }

    void removeAllByPostalAddressCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 0;

        printCommandResult(
                status,
                Localization.get("message.organizations_by_postal_address_removed"),
                ApplicationErrorMessages.removeAllByPostalAddressCommandGetErrorMessage(error)
        );
    }

    void modifyOrganizationCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 0;

        printCommandResult(
                status,
                Localization.get("message.organization_modified"),
                ApplicationErrorMessages.modifyOrganizationGetErrorMessage(error)
        );
    }

    void maxByFullNameCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert (result.length == 1 && result[0] instanceof Organization) || error != null;

        printCommandResult(
                status,
                status == ExecutionStatus.SUCCESS ? ((Organization) result[0]).toYaml() : ""
        );
    }

    void sumOfAnnualTurnoverCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert (result.length == 1 && result[0] instanceof Float) || error != null;

        printCommandResult(
                status,
                status == ExecutionStatus.SUCCESS ?
                        String.format("%s: %f.", Localization.get("message.sum_of_annual_turnover"), (Float) result[0])
                        : ""
        );
    }

    void executeScriptCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 1;
        assert result[0] instanceof String;

        String filename = (String) result[0];

        printCommandResult(
                status,
                Localization.get("message.script_execution.started"),
                ApplicationErrorMessages.executeScriptCommandGetErrorMessage(error, filename)
        );
    }

    void showHistoryCommandCallback(ExecutionStatus status, Exception error, Object... result) {
        assert result.length == 1;
        assert result[0] instanceof CircledStorage;

        CircledStorage<String> history = (CircledStorage<String>) result[0];
        StringBuilder stringBuilder = new StringBuilder();
        history.applyFunctionOnAllElements((String s) -> stringBuilder.append(s).append("\n"));

        printCommandResult(
                status,
                stringBuilder.toString()
        );
    }

    private void loadCommands() {
        commandsWithLocaleNames = new HashMap<>();

        for (ImmutablePair<String, Command> command : commandList) {
            commandsWithLocaleNames.put(Localization.get(command.first()), command.second());
        }
    }

    private void localize() throws IOException {
        Localization.askUserForALanguage(bufferedReaderWithQueueOfStreams);
        loadCommands();
    }

    private void processCommand(String command) {
        String[] allArguments = command.split(" +");
        String commandName = allArguments[0];
        String[] args = Arrays.copyOfRange(allArguments, 1, allArguments.length);

        if (commandsWithLocaleNames.containsKey(commandName)) {
            commandsWithLocaleNames.get(commandName).execute(args);
            addCommandToHistory(commandName);
        } else {
            System.out.println(Localization.get("message.command.not_found"));
        }
    }

    private static void printIntroductionMessage() {
        System.out.println(Localization.get("message.introduction"));
    }

    public void addCommandToHistory(String command) {
        commandsHistory.add(command);
    }
}

File: src/application/Application.kt
package application

import commands.client.*
import commands.client.core.Command
import database.DatabaseInterface
import kotlinx.coroutines.*
import lib.BufferedReaderWithQueueOfStreams
import lib.Localization
import lib.collections.CircledStorage
import network.client.DatabaseCommand
import java.io.InputStreamReader

class Application(val database: DatabaseInterface, dispatcher: CoroutineDispatcher) {
    private val applicationScope = CoroutineScope(dispatcher)
    val commandsHistory: CircledStorage<String> = CircledStorage(11)
    val bufferedReaderWithQueueOfStreams: BufferedReaderWithQueueOfStreams = BufferedReaderWithQueueOfStreams(
        InputStreamReader(System.`in`)
    )

    private var running = false
    private var localNameToDatabaseCommand: HashMap<String, DatabaseCommand> = HashMap()

    private val databaseCommandToExecutor: Map<DatabaseCommand, Command> = mapOf(
        DatabaseCommand.HELP to HelpCommand(),
        DatabaseCommand.INFO to InfoCommand(database),
        DatabaseCommand.SHOW to ShowCommand(database),
        DatabaseCommand.ADD to AddCommand(database),
        DatabaseCommand.UPDATE to UpdateCommand(database),
        DatabaseCommand.REMOVE_BY_ID to RemoveByIdCommand(database),
        DatabaseCommand.CLEAR to ClearCommand(database),
        DatabaseCommand.SAVE to SaveCommand(database),
        DatabaseCommand.READ to ReadCommand(database),
        DatabaseCommand.EXECUTE_SCRIPT to ExecuteScriptCommand(this),
        DatabaseCommand.EXIT to ExitCommand(this),
        DatabaseCommand.REMOVE_HEAD to RemoveHeadCommand(database),
        DatabaseCommand.ADD_IF_MAX to AddIfMaxCommand(database),
        DatabaseCommand.HISTORY to PrintHistoryCommand(this),
        DatabaseCommand.MAX_BY_FULL_NAME to MaxByFullNameCommand(database),
        DatabaseCommand.REMOVE_ALL_BY_POSTAL_ADDRESS to RemoveAllByPostalAddressCommand(database),
        DatabaseCommand.SUM_OF_ANNUAL_TURNOVER to SumOfAnnualTurnoverCommand(database)
    )

    private val argumentForCommand: Map<DatabaseCommand, (String?) -> Any?> = mapOf(
        DatabaseCommand.HELP to { null },
        DatabaseCommand.INFO to { null },
        DatabaseCommand.CLEAR to { null },
        DatabaseCommand.SAVE to { null },
        DatabaseCommand.EXIT to { null },
        DatabaseCommand.REMOVE_HEAD to { null },
        DatabaseCommand.HISTORY to { null },
        DatabaseCommand.MAX_BY_FULL_NAME to { null },
        DatabaseCommand.SUM_OF_ANNUAL_TURNOVER to { null },
        DatabaseCommand.SHOW to { it },
        DatabaseCommand.READ to { it },
        DatabaseCommand.EXECUTE_SCRIPT to { it },
        DatabaseCommand.REMOVE_BY_ID to { it?.toIntOrNull() },
        DatabaseCommand.ADD to {
            OrganizationBuilder.constructOrganization(
                bufferedReaderWithQueueOfStreams,
                false
            )
        },
        DatabaseCommand.UPDATE to {
            OrganizationBuilder.constructOrganization(
                bufferedReaderWithQueueOfStreams,
                true
            )
        },
        DatabaseCommand.REMOVE_ALL_BY_POSTAL_ADDRESS to {
            OrganizationBuilder.constructAddress(
                bufferedReaderWithQueueOfStreams,
                false
            )
        },
        DatabaseCommand.ADD_IF_MAX to {
            OrganizationBuilder.constructOrganization(
                bufferedReaderWithQueueOfStreams,
                false
            )
        }
    )

    private fun loadCommands() {
        localNameToDatabaseCommand.clear()

        for ((key, value) in commandNameToDatabaseCommand) {
            localNameToDatabaseCommand[Localization.get(key)] = value
        }
    }

    private fun localize() {
        Localization.askUserForALanguage(bufferedReaderWithQueueOfStreams)
        loadCommands()
    }

    fun start() = runBlocking {
        localize()
        printIntroductionMessage()
        running = true

        while (running) {
            val line = bufferedReaderWithQueueOfStreams.readLine()
            processCommand(line.trim())
        }
    }

    fun stop() {
        running = false
    }

    private suspend fun processCommand(input: String) {
        val allArguments = splitInputIntoArguments(input)

        if (allArguments.isEmpty()) {
            return
        }

        val commandName = allArguments[0]
        val argument = allArguments.getOrNull(1)
        val databaseCommand = localNameToDatabaseCommand[commandName]

        if (databaseCommand == null) {
            System.out.printf(Localization.get("message.command.not_found"), commandName)
            return
        }

        val executionArgument = argumentForCommand[databaseCommand]!!.invoke(argument)
        applicationScope.launch { executeCommand(databaseCommand, executionArgument) }
    }

    private suspend fun executeCommand(databaseCommand: DatabaseCommand, argument: Any?) {
        val executor = databaseCommandToExecutor[databaseCommand]
        val result = executor!!.execute(argument)

        if (result.isSuccess) {
            val successMessage = commandSuccessMessage(databaseCommand, result.getOrNull())
            println(successMessage)
        } else {
            val exception = result.exceptionOrNull()!!
            val errorMessage = exceptionToMessage(exception)
            println(errorMessage)
        }

        addCommandToHistory(databaseCommand.name)
    }

    private fun addCommandToHistory(command: String) {
        commandsHistory.add(command)
    }

    private fun printIntroductionMessage() {
        println(Localization.get("message.introduction"))
    }
}
